[6.15] Switch ALL Novatek nt36672a based panels to common driver

minlexx opened the pull request:
Start with lavender. Later add jasmine, tulip, maybe something else?

Fixes #70

minlexx committed:
drm/panel: nt36672a: Switch from 2-byte arrays to init functions

The driver was made with extensibility in mind, but it is not enough.
In preparation for extending the driver with support for more panels,
stop using arrays of nt36672a_panel_cmd, which incorrectly assumed
that all panels have init sequences consisting of 2 bytes. This is
not true for all panels using nt36672a controller. Instead, use
dedicated init and shutdown functions for each panel. This allows
for more flexibility, solving at least 2 problems:

 * different DSI commands length for different panels
 * different delays for msleep() calls for different panels

Make it so that each panel can initialize the way it prefers.

Combine init sequences in .on_cmds_1 and .on_cmds_2 into a new init
finction, and code from .off_cmds to new off_cmds function. Struct
nt36672a_panel_cmd and function nt36672a_send_cmds() are no longer
needed, so remove them.

While doing this, also switch to modern multi-api for sending
commands, which greatly simplifies error handling using accumulated
error counter.

This refactoring change should have no functional effect on existing
devices.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/panel: nt36672a: Add Tianma panel used in Xiaomi Redmi Note 7

Add support for Tianma TL063FVMCA01-00 video mode panel based on
nt36672a, used in Xiaomi Redmi Note 7 smartphone (xiaomi-lavender).

Tianma panel is one of the 3 panel variants that you can find inside
xiaomi-lavender.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
fixup! arm64: dts: qcom: sdm660-xiaomi-lavender: Split by display

use touch in follower mode

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/panel: nt36672a: Add Tianma panel used in Xiaomi Mi A2

Add support for Tianma TL060FVXS16-00 video mode panel based on
Novatek nt36672a, used in Xiaomi Mi A2 smartphone (xiaomi-jasmine).
It shares some code with already supported panel in this driver
for xiaomi-lavender, so we can reuse the same panel off function.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/panel: nt36672a: Add Tianma panel used in Xiaomi Redmi Note 6 Pro

Add support for Tianma TL063FVMC43-02 video mode panel based on
Novatek nt36672a, used in Xiaomi Redmi Note 6 Pro smartphone
(xiaomi-tulip). It shares almost exactly the same init/off
sequences with already supported panel in this driver for Xiaomi
Redmi Note 7 (xiaomi-lavender).

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

barni2000 commented on arch/arm64/boot/dts/qcom/sdm660-xiaomi-lavender-tianma.dts:
```diff
@@ -15,24 +15,27 @@
+		compatible = "novatek,nt36672a-ts";
+		reg = <0x1>;
+	
+		iovcc-supply = <&vreg_l11a_1p8>;
```
```suggestion
		vddio-supply = <&vreg_l11a_1p8>;
```

minlexx replied:
As far as I can tell it's called iovcc in https://elixir.bootlin.com/linux/v6.15/source/drivers/input/touchscreen/novatek-nvt-ts.c#L226

barni2000 replied:
Oh, sorry i have mixed it with the panel driver :)

barni2000 reviewed (dismissed):
this driver naming the iovcc supply as vddio

minlexx committed:
fixup! arm64: dts: qcom: sdm660: Add device tree for Xiaomi Mi A2 (jasmine)

use touch in follower mode

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
fixup! arm64: dts: qcom: add device-tree for Redmi Note 6 Pro (tulip)

minlexx committed:
Input: novatek-nvt-ts - support being panel follower

Sometimes Novatek touchscreen is paired together with Novatek panel,
and they both need to be powered together in sync.

Add support for nvt-ts driver to operate in panel follower mode, in
which touchscreen isn't in full control of its lifecycle.
This is managed using callbacks from panel, to get informed when the
panel is powered on and off. From there we can match the nvt-ts
device's power state to that of the panel.

Without this probing touchscreen before panel sometimes randomly fails
with I2C read errors.

 * split some code from probe function to separate initial_power_on
   function, which is called once from panel_prepared callback, or
   from probe function, depending on mode;
 * update Kconfig to depend on DRM.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

iAboothahir committed:
fixup! arm64: dts: qcom: Add initial support for asus-x00td

iAboothahir committed:
drm/panel: nt36672a: Add TXD panel used in Asus Zenfone Max Pro M1

Add support for TXD TXDI600YANPA-43V3 video mode panel based on
Novatek nt36672a, used in Asus Zenfone Max Pro M1 smartphone
(asus-x00td).

Signed-off-by: iAboothahir <aboothahirpkd@gmail.com>

minlexx committed:
drm/panel: nt36672a: Switch to modern refcounted allocation API

Use the new refcounted allocation API devm_drm_panel_alloc() to allocate
the panel in place of devm_kzalloc().

Remove call to drm_panel_init() because it's part of devm_drm_panel_alloc().

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
drm/panel: nt36672a: Change driver name to novatek for consistency

Since this driver should represent not only devices manufactured by
Tianma Micro-electronics, but also from other companies, it makes more
sense to name it according to what driver file name already is: novatek.

This does not change neither module file name nor Kconfig symbol, only
driver identification in dmesg. This makes it possible to e.g. grep
dmesg for "novatek" and get messages from both nt36672a-based panel and
Novatek touchscreen.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

flamingradian committed:
drm/panel: nt36672a: Add Shenchao panel used in Xiaomi Redmi Note 7

Add support for the Shenchao panel used in the Xiaomi Redmi Note 7
(xiaomi-lavender) smartphone. It is one of the 3 display panels found in
xiaomi-lavender.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>

flamingradian committed:
dt-bindings: arm: qcom: Add xiaomi-lavender with shenchao panel

Add the compatible for the Xiaomi Redmi Note 7 with the Shenchao display
panel.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>

flamingradian committed:
arm64: dts: qcom: add Xiaomi Redmi Note 7 with Shenchao panel

The Redmi Note 7 can have 3 different display panels. Add support for
the Shenchao panel.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>

barni2000 commented on drivers/gpu/drm/panel/panel-novatek-nt36672a.c:
```diff
@@ -71,6 +71,66 @@ static inline struct nt36672a_panel *to_nt36672a_panel(struct drm_panel *panel)
+	return dsi_ctx.accum_err;
+}
+
+static int shenchao_lavender_off_cmds(struct drm_panel *panel)
```
You can use tianma_lavender_off_cmds because the both function are completely same.

minlexx replied:
Fixed this one

minlexx commented:
Rebased on target branch after few MRs merged into it

minlexx committed:
HACK drm/panel: move some dsi commands from unprepare to disable

The commit 007ac026 ("drm/msm/dsi: switch to DRM_PANEL_BRIDGE")
breaks panels which send DSI commands in their .unprepare callbacks.
Migrate to using .disable for that for some panels.

Inspired-by: https://gitlab.com/sdm845-mainline/linux/-/commit/88d6f2060e1f4fd283075079088844e0474e8d8b

Co-developed-by: Joel Selvaraj <joelselvaraj.oss@gmail.com>
Signed-off-by: Joel Selvaraj <joelselvaraj.oss@gmail.com>
Co-developed-by: Caleb Connolly <caleb.connolly@linaro.org>
Signed-off-by: Caleb Connolly <caleb.connolly@linaro.org>
Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
dt-bindings: display: novatek,nt36672a: Add 5 new panels

The driver now supports 5 new panels for 4 different smartphones.
Describe them.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
dt-bindings: vendor-prefixes: Add Shenzhen TXD Technology Co., Ltd.

Add prefix for Shenzhen TXD Technology Co., Ltd., manufacturer of
medium sized LCD and camera modules for mobile phones, tablets and
automotive.

[1] https://en.txdkj.com/

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
dt-bindings: vendor-prefixes: Add Shenzhen Zhuohong Electronic

Shenzhen Zhuohong Electronic Technology Co., Ltd. is a manufacturer
of LCD/IPS displays based in Shenzhen, Mainland China.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx commented:
Continued for kernel 6.16 in #114
